File: app/channels/ping_channel.rb
class PingChannel < ApplicationCable::Channel
  STREAM = "mefkit_ping_channel"

  def subscribed
    Rails.logger.debug { "PingChannel subscribed" }
    stream_from STREAM
    ensure_confirmation_sent
    transmit(
      ping: {
        msg: "subscribed",
      },
    )
  end

  def unsubscribed
    Rails.logger.debug { "PingChannel unsubscribed" }
    # Any cleanup needed when channel is unsubscribed
  end

  def pong(data)
    Rails.logger.info { ">>>>>> PingChannel pong: #{data}" }
    transmit(
      ping: {
        msg: "ACK: #{data["msg"]}",
      },
    )
  end

  class << self
    def ping(msg)
      ActionCable.server.broadcast(
        STREAM,
        ping: {
          msg: msg,
        },
      )
    end
  end
end

File: app/javascript/channels/ping_channel.js
import consumer from "./consumer";

let callbacks = new Set();

const pingChannel = consumer.subscriptions.create("PingChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
    console.log("PingChannel connected");
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
    console.error("PingChannel disconnected");
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    console.log("PingChannel received", data);
    for (let callback of callbacks) {
      callback.call(null, data.ping.msg);
    }
  }
});

const pong = msg => {
  pingChannel.perform("pong", { msg });
};

const receiveMessage = fn => {
  callbacks.add(fn);
};

export { pong, receiveMessage };
